### PR TITLE
[WIP] hooks: make hooks dependent on available utils

### PR DESCRIPTION
**Pull request by phlax**
Signed-off-by: Ryan Northey <ryan@synca.io>

Commit Message: hooks: skip code_format.py if $BUILDIFER_PATH is not set
Additional Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Comment by phlax** [comment]
cc @mattklein123 

i can type `--no-verify` almost as fast i can type `import pdb; pdb.set_trace()` `8/`

it would be great to be able to push without having to worry about push hooks

this will make it much easier to add separate hooks for linting tools

File: support/hooks/pre-push [hunk 1]
```diff
@@ -58,12 +58,18 @@ do
...
+        echo -ne "  Checking format for $i - "
+        "$SCRIPT_DIR"/code_format/check_format.py check $i
+        if [[ $? -ne 0 ]]; then
+          exit 1
+        fi
+      else
+        echo -e "\nMissing \$BUILDIFIER_PATH, skipping code_format/check_format.py\n" >&2
       fi
 
+      ./tools/code_format/glint.sh "$i"
```
Comment (phlax) [inline support/hooks/pre-push:71]: remove for now

**Comment by phlax** [comment]
nm - i think i opened this in frustration

ill make it a bag of hooks and maybe some cleanup to the hook file

File: support/hooks/pre-push [hunk 2]
```diff
@@ -13,81 +13,189 @@ DUMMY_SHA=0000000000000000000000000000000000000000
...
-    if [[ $? -ne 0 ]]; then
-      exit 1
+check_shellcheck () {
+    local file="$1" isshell=
+    if [[ -n "$SKIP_SHELLCHECK" ]]; then
+        return
+    fi
+    if [[ ${file: -3} == ".sh" ]]; then
+        isshell=yes
+    elif grep "^#!/bin/bash" "$file"; then
```
Comment (phlax) [inline support/hooks/pre-push:126]: merge these to a single expression

File: tools/code_format/glint.sh [hunk 3]
```diff
@@ -52,15 +53,25 @@ find_text_files () {
...
 #
 # note: we may want to use python if this grows in complexity
 #
-for file in $(find_text_files); do
-    check_new_line "$file"
-    check_mixed_tabs_spaces "$file"
-    check_trailing_whitespace "$file"
-done
+if [[ -n "$FILEPATH" ]]; then
+    echo "RUN 11111"
```
Comment (phlax) [inline tools/code_format/glint.sh:67]: remove

File: tools/code_format/format_python_tools.py [hunk 4]
```diff
@@ -28,7 +28,7 @@ def collect_files():
     return matches
 
 
-def validate_format(fix=False):
+def validate_format(fix=False, path=None):
```
Comment (phlax) [inline tools/code_format/format_python_tools.py:31]: remove todo above

File: support/hooks/pre-push [hunk 5]
```diff
@@ -13,81 +13,189 @@ DUMMY_SHA=0000000000000000000000000000000000000000
...
+# skip message/logic for code_format
+SKIP_CODE_FORMAT="${SKIP_CODE_FORMAT:-}"
+SKIP_CODE_FORMAT_PROBLEM="WARNING: Missing \$BUILDIFIER_PATH or \$CLANG_FORMAT, skipping code_format/check_format.py"
+SKIP_CODE_FORMAT_SOLUTION="See https://github.com/envoyproxy/envoy/blob/main/bazel/README.md#running-clang-format-without-docker for further information"
+
+# skip message/logic for autoformat
+SKIP_PYTHON_AUTOFORMAT="${SKIP_PYTHON_AUTOFORMAT:-}"
+
+# skip message/logic for flake8
+SKIP_FLAKE8="${SKIP_FLAKE8:-}"
```
Comment (phlax) [inline support/hooks/pre-push:29]: add problem/solutions for these

File: support/hooks/pre-push [hunk 6]
```diff
@@ -13,81 +13,189 @@ DUMMY_SHA=0000000000000000000000000000000000000000
...
+# skip logic for protos
+SKIP_PROTOS="${SKIP_PROTOS:-}"
+
+# NOTE: The `tools` directory will be the same relative to the support
+# directory, independent of whether we're in a submodule, so no need to use
+# our `relpath` helper.
+SRC_DIR="$(dirname "$(realpath "$0")")/../../"
+SCRIPT_DIR="${SRC_DIR}/tools"
+
+CHECKS=(
```
Comment (phlax) [inline support/hooks/pre-push:43]: move on up

File: support/hooks/pre-push [hunk 7]
```diff
@@ -13,81 +13,189 @@ DUMMY_SHA=0000000000000000000000000000000000000000
...
+        else
+            # Updating branch. Verify new commits.
+            RANGE="$REMOTE_SHA..$LOCAL_SHA"
+        fi
+
+        check_dco "$RANGE"
+        for i in $(git diff --name-only "$RANGE" --diff-filter=ACMR --ignore-submodules=all 2>&1); do
+            for checkinfo in "${CHECKS[@]}"; do
+                read -ra check <<< "${checkinfo//:/ }"
+                if echo "$i" | grep -E "${check[1]}"; then
```
Comment (phlax) [inline support/hooks/pre-push:195]: add an explanation as to what this does

File: support/hooks/pre-push [hunk 8]
```diff
@@ -13,81 +13,189 @@ DUMMY_SHA=0000000000000000000000000000000000000000
...
+    if [[ -n "$FLAKE8_PATH" ]];  then
+        "$FLAKE8_PATH" "$i" || exit 1
+    else
+        echo -e "\n  $FLAKE8_PROBLEM\n" >&2
+        echo -e "\n  $FLAKE8_SOLUTION\n" >&2
+        SKIP_FLAKE8=yes
+    fi
+}
 
-    # TODO(mattklein123): Optimally we would be able to do this on a per-file basis.
```
Comment (phlax) [inline support/hooks/pre-push:74]: bring back comment

File: tools/code_format/format_python_tools.py [hunk 9]
```diff
@@ -14,7 +14,6 @@ def collect_files():
     Returns: A collection of python files in the tools directory excluding
        any directories in the EXCLUDE_LIST constant.
     """
-    # TODO: Add ability to collect a specific file or files.
```
Comment (phlax) [inline tools/code_format/format_python_tools.py:17]: this is resolved at least for the case of a single file - this method is then avoided

File: tools/code_format/format_python_tools.py [hunk 10]
```diff
@@ -70,6 +70,8 @@ def display_fix_results(successful_files, failed_files):
     parser = argparse.ArgumentParser(description='Tool to format python files.')
     parser.add_argument(
         'action', choices=['check', 'fix'], default='check', help='Fix invalid syntax in files.')
+    parser.add_argument(
+        'path', default=None, help='Fix invalid syntax in files.')
```
Comment (phlax) [inline tools/code_format/format_python_tools.py:74]: wrong help

File: support/hooks/pre-push [hunk 11]
```diff
@@ -11,83 +11,207 @@
...
-    "$SCRIPT_DIR"/code_format/format_python_tools.sh check
-    if [[ $? -ne 0 ]]; then
-      exit 1
+check_python_autoformat () {
+    # Run python autoformatter
+    local file="$1"
+    if [[ -n "$SKIP_PYTHON_AUTOFORMAT" ]]; then
+        return
     fi
+    "$SCRIPT_DIR"/code_format/format_python_tools.sh check "$file" || :
```
Comment (phlax) [inline support/hooks/pre-push:171]: this is failing for me

File: support/hooks/pre-push [hunk 12]
```diff
@@ -11,83 +11,207 @@
...
 
-    "$SCRIPT_DIR"/code_format/format_python_tools.sh check
-    if [[ $? -ne 0 ]]; then
-      exit 1
+check_python_autoformat () {
+    # Run python autoformatter
+    local file="$1"
+    if [[ -n "$SKIP_PYTHON_AUTOFORMAT" ]]; then
+        return
     fi
```
Comment (phlax) [inline support/hooks/pre-push:170]: add debug message

File: support/hooks/pre-push [hunk 13]
```diff
@@ -11,83 +11,207 @@
...
-      # (though there is no way to know for sure). In the extremely uncommon case in which someone
-      # pushes more than 1 new commit to a branch, CI will enforce full checking.
-      RANGE="$LOCAL_SHA~1..$LOCAL_SHA"
-    else
-      # Updating branch. Verify new commits.
-      RANGE="$REMOTE_SHA..$LOCAL_SHA"
-    fi
+echo "Running pre-push check; to skip this step use 'push --no-verify' or check support/hooks/README.md for more info"
+
+CHECKS=(
```
Comment (phlax) [inline support/hooks/pre-push:16]: add explanation

File: support/hooks/pre-push [hunk 14]
```diff
@@ -11,83 +11,207 @@
...
+    "glint:*"
+    "shellcheck:*"
+    "flake8:*.py"
+    "python_autoformat:*.py"
+    "spelling:*"
+    "protos:*.proto"
+    "repositories:*.bzl")
+
+DEBUG="${DEBUG_COMMIT_PUSH:-}"
+FLAKE8_PATH=${FLAKE_PATH:-$(which flake8)}
```
Comment (phlax) [inline support/hooks/pre-push:27]: move these down

File: tools/code_format/format_python_tools.sh [hunk 15]
```diff
@@ -6,10 +6,13 @@
 
 . tools/shell_utils.sh
 
+COMMAND="$1"
+PATH="${2:-}"
```
Comment (phlax) [inline tools/code_format/format_python_tools.sh:10]: doh

File: tools/code_format/BUILD [hunk 16]
```diff
@@ -10,13 +10,11 @@ exports_files([
     "check_format.py",
     "header_order.py",
     "envoy_build_fixer.py",
-    "flake8.conf",
 ])
 
 py_binary(
     name = "python_flake8",
     srcs = ["python_flake8.py"],
-    data = [":flake8.conf"],
```
Comment (phlax) [inline tools/code_format/BUILD:19]: this file is moved to the root of the repo so it is automatically picked up running the tool

File: tools/code_format/format_python_tools.py [hunk 17]
```diff
@@ -67,9 +67,12 @@ def display_fix_results(successful_files, failed_files):
 
 
 if __name__ == '__main__':
+    # py change
```
Comment (phlax) [inline tools/code_format/format_python_tools.py:70]: remove

File: .flake8 [hunk 18]
```diff
@@ -2,7 +2,7 @@
 [flake8]
 
 # TODO(phlax): make this an exclusive list and enable most tests
-select = N802,F401,E111,F821
+select = N802,F401,E111,F821,E9,F63,F72,F82
```
Comment (phlax) [inline .flake8:5]: this merges the checks that are being run in 2 cases of flake8 in ci

File: support/hooks/pre-push [hunk 19]
```diff
@@ -11,83 +11,273 @@
...
+SHELLCHECK_PATH=${FLAKE_PATH:-$(which shellcheck)}
+SHELLCHECK_PROBLEM="WARNING: Missing flake8 command in \$PATH"
+SHELLCHECK_SOLUTION="Install shellcheck for your system."
+SKIP_SHELLCHECK="${SKIP_SHELLCHECK:-}"
+
+# skip logic for protos
+SKIP_PROTOS="${SKIP_PROTOS:-}"
+
+# skip message/logic for yapf
+YAPF_PATH=${FLAKE_PATH:-$(which yapf)}
```
Comment (phlax) [inline support/hooks/pre-push:62]: wrong path

File: support/hooks/pre-push [hunk 20]
```diff
@@ -11,83 +11,273 @@
...
+SKIP_PYTHON_AUTOFORMAT="${SKIP_PYTHON_AUTOFORMAT:-}"
+
+# skip message/logic for flake8
+FLAKE8_PATH=${FLAKE_PATH:-$(which flake8)}
+FLAKE8_PROBLEM="WARNING: Missing flake8 command in \$PATH"
+FLAKE8_SOLUTION="Install flake8 using pip."
+SKIP_FLAKE8="${SKIP_FLAKE8:-}"
+
+# skip message/logic for shellcheck
+SHELLCHECK_PATH=${FLAKE_PATH:-$(which shellcheck)}
```
Comment (phlax) [inline support/hooks/pre-push:53]: wrong path

File: support/hooks/pre-push [hunk 21]
```diff
@@ -11,83 +11,273 @@
...
+        log_error yapf "$YAPF_PROBLEM" "    $YAPF_SOLUTION"
+        handle_missing yapf
+    fi
+}
+
+check_glint () {
+    # Run glint on all files to ensure general formatting compliance
+    local file="$1"
+    log_check glint "$file"
+    "$SCRIPT_DIR"/code_format/glint.sh "$1"
```
Comment (phlax) [inline support/hooks/pre-push:180]: `|| handle_error ...`

File: tools/shell_utils.sh [hunk 22]
```diff
@@ -25,5 +25,5 @@ python_venv() {
   pip3 install -r "${SCRIPT_DIR}"/requirements.txt
 
   shift
-  python3 "${SCRIPT_DIR}/${PY_NAME}.py" "$*"
+  python3 "${SCRIPT_DIR}/${PY_NAME}.py" "${@}"
```
Comment (phlax) [inline tools/shell_utils.sh:28]: this isnt strictly necessary for this pr as i switched the hook to using the python file directly

i still think this is a good idea tho, so included it.

**Comment by phlax** [comment]
@mattklein123 i have updated the git hook so we can run more checks etc and added a README to explain how it works - would  you PTAL, thanks

it duplicates the readme in `/support` already - altho, i think that may be not up to date - not sure

i can integrate/update/replace as required after

in terms of the hook, with this PR it will warn if a need tool for a particular check is not present - but not fail

it also doesnt auto-install any utilities, which imho is preferable

anyhow - i think this works pretty well and will give better feedback to users when they need to install something or fix something

afaict it autofixes the same as the existing hook - but not certain - it needs some further testing before it is ready 

it also adds some required and related fixes - the main thing atm tho is to check out the readme in terms of general direction

File: support/hooks/README.md [hunk 23]
```diff
@@ -0,0 +1,138 @@
+
+### The git push hook
+
```
Comment (phlax) [inline support/hooks/README.md:3]: readme is here

File: support/hooks/pre-push [hunk 24]
```diff
@@ -11,83 +11,267 @@
...
+        isshell=yes
+    fi
+    if [[ -z "$isshell" ]]; then
+        return
+    fi
+    log_check shellcheck "$file"
+    if [[ -n "$SHELLCHECK_PATH" ]];  then
+        "$SHELLCHECK_PATH" -x "$file" || handle_error shellcheck
+    else
+        log_error shellcheck "$SHELLCHECK_PROBLEM" "$SHELLCHECK_SOLUTION"
```
Comment (phlax) [inline support/hooks/pre-push:164]: this is wrong

File: support/hooks/pre-push [hunk 25]
```diff
@@ -11,83 +11,267 @@
...
+}
 
+check_protos () {
+    # Run proto_format.sh
+    # it only runs once and only if a .proto file is found to be changed
+    if [[ -n "$SKIP_PROTOS" ]]; then
+        return
+    fi
+    log_check protos "$file"
+    "$SCRIPT_DIR"/proto_format/proto_format.sh check || handle_error protos
```
Comment (phlax) [inline support/hooks/pre-push:140]: check what is called/needed here

File: support/hooks/pre-push [hunk 26]
```diff
@@ -11,83 +11,267 @@
...
+        "$SHELLCHECK_PATH" -x "$file" || handle_error shellcheck
+    else
+        log_error shellcheck "$SHELLCHECK_PROBLEM" "$SHELLCHECK_SOLUTION"
+        SKIP_SHELLCHECK=yes
+        handle_missing yapf "$YAPF_PROBLEM" "    $YAPF_SOLUTION"
+    fi
+}
+
+check_glint () {
+    # Run glint on all files to ensure general formatting compliance
```
Comment (phlax) [inline support/hooks/pre-push:171]: improve comment/s

File: tools/code_format/glint.sh [hunk 27]
```diff
@@ -12,6 +12,7 @@ ERRORS=
 MISSING_NEWLINE=0
 MIXED_TABS_AND_SPACES=0
 TRAILING_WHITESPACE=0
+FILEPATH="${1}"
```
Comment (phlax) [inline tools/code_format/glint.sh:15]: `{1:-}`

File: tools/code_format/format_python_tools.sh [hunk 28]
```diff
@@ -6,10 +6,13 @@
 
 . tools/shell_utils.sh
 
+FORMAT_ACTION="$1"
+FORMAT_PATH="${2:-}"
```
Comment (phlax) [inline tools/code_format/format_python_tools.sh:10]: this is not required for this pr, at least

File: support/hooks/README.md [hunk 29]
```diff
@@ -0,0 +1,138 @@
...
+...
+error: failed to push some refs to 'github.com:phlax/envoy'
+
+```
+
+#### Debugging hooks
+
+Mostly the hooks should be quiet (ideally at least!), unless there is a problem.
+
+If you wish to see what is being run set the debug flag in `.envoypush`
```
Comment (phlax) [inline support/hooks/README.md:95]: `DEBUG`

File: support/hooks/README.md [hunk 30]
```diff
@@ -0,0 +1,138 @@
...
+
+```console
+$ echo "FAIL_ON_MISSING=yes" >> .envoypush
+$ git push
+Running pre-push check; to skip this step use 'push --no-verify' or check support/hooks/README.md for more info
+[check:code_format] WARNING: Missing $BUILDIFIER_PATH or $CLANG_FORMAT, skipping code_format/check_format.py
+[check:code_format]     See https://github.com/envoyproxy/envoy/blob/main/bazel/README.md#running-clang-format-without-docker for further information
+[check:code_format] ERROR: missing util
+...
+error: failed to push some refs to 'github.com:phlax/envoy'
```
Comment (phlax) [inline support/hooks/README.md:87]: user/envoy

**Comment by mattklein123** [comment]
Is this still WIP or is it ready for review? LMK when ready.

**Comment by phlax** [comment]
its still wip - but i wanted to draw your attention to the readme here https://github.com/envoyproxy/envoy/blob/80d584ce2837e72a6631ee42bcdbb48315fdcbdb/support/hooks/README.md

File: support/hooks/pre-push [hunk 31]
```diff
@@ -11,83 +11,280 @@
...
-      echo "  Checking spelling for $i"
-      "$SCRIPT_DIR"/spelling/check_spelling_pedantic.py check $i
-      if [[ $? -ne 0 ]]; then
-        exit 1
-      fi
-    done
+    log_check dco
+    signed_off=$(git rev-list --no-merges --grep "^Signed-off-by: " "$range")
+    not_signed_off=$(git rev-list --no-merges "$range" | grep -Fxv "$signed_off")
+    if [ -n "$not_signed_off" ]; then
```
Comment (phlax) [inline support/hooks/pre-push:100]: brackets

File: .gitignore [hunk 32]
```diff
@@ -37,3 +38,4 @@ cmake-build-debug
 /linux
 bazel.output.txt
 *~
+xyz
```
Comment (phlax) [inline .gitignore:41]: rm

File: support/bootstrap [hunk 33]
```diff
@@ -20,9 +20,9 @@ fi
 # commit 82b040a60561cf94dec3197ea88ae15e57bcaa97, which also carries the Apache
 # V2 license, and has deployed this code successfully for some time.
 abspath() {
-    cd "$(dirname "${1}")"
+    cd "$(dirname "${1}")" || return 1
```
Comment (phlax) [inline support/bootstrap:23]: this is shellcheck cleanups

File: tools/code_format/format_python_tools.sh (hunk 28 above)
Comment (phlax) [reply tools/code_format/format_python_tools.sh:10]: actually its  worth keeping

**Comment by phlax** [comment]
there are no api changes, i was testing proto changes

File: support/hooks/pre-push [hunk 34]
```diff
@@ -11,83 +11,314 @@
...
-      # New branch. Verify the last commit, since this is very likely where the new code is
-      # (though there is no way to know for sure). In the extremely uncommon case in which someone
-      # pushes more than 1 new commit to a branch, CI will enforce full checking.
-      RANGE="$LOCAL_SHA~1..$LOCAL_SHA"
-    else
-      # Updating branch. Verify new commits.
-      RANGE="$REMOTE_SHA..$LOCAL_SHA"
-    fi
+# Set non-empty value to quieten push hook
+SKIP_PUSH_MESSAGE=
```
Comment (phlax) [inline support/hooks/pre-push:15]: get from env

File: support/hooks/pre-push [hunk 35]
```diff
@@ -11,83 +11,314 @@
...
+EXIT_ON_ERROR="${EXIT_ON_ERROR:-}"
+FAIL_ON_MISSING="${FAIL_ON_MISSING:-}"
+PUSH_ON_FAILURE="${PUSH_ON_FAILURE:-}"
+FAILED=
+
+# DCO
+DCO_ERROR_COMMITS=(
+    "ERROR: The following commits do not have DCO signoff:"
+    ""
+    "See here for information on how to fix")
```
Comment (phlax) [inline support/hooks/pre-push:53]: add url

File: support/hooks/pre-push [hunk 36]
```diff
@@ -11,83 +11,314 @@
...
+    ""
+    "    Install flake8 using pip.")
+SKIP_FLAKE8="${SKIP_FLAKE8:-}"
+
+# skip message/logic for shellcheck
+SHELLCHECK_PATH=${SHELLCHECK_PATH:-$(command -v shellcheck)}
+SHELLCHECK_WARNING=(
+    "WARNING: Missing shellcheck command in \$PATH"
+    ""
+    "    Install shellcheck for your system.")
```
Comment (phlax) [inline support/hooks/pre-push:75]: add url

File: support/hooks/pre-push [hunk 37]
```diff
@@ -11,83 +11,314 @@
...
+SKIP_SHELLCHECK="${SKIP_SHELLCHECK:-}"
+
+# skip logic for protos - requires bazel
+SKIP_PROTOS="${SKIP_PROTOS:-}"
+BAZEL_PATH=${BAZEL_PATH:-$(command -v bazel)}
+PROTOS_WARNING=(
+    "WARNING: Missing bazel command in \$PATH"
+    ""
+    "    Install bazel for your system.")
```
Comment (phlax) [inline support/hooks/pre-push:84]: add url

File: support/hooks/pre-push [hunk 38]
```diff
@@ -11,83 +11,314 @@
...
-  fi
-done
+}
 
-exit 0
+handle_error () {
+    local failure="$1"
+    FAILED=yes
+    if [[ -n "$EXIT_ON_ERROR" ]]; then
+        log_error "$failure" "FAILED"
```
Comment (phlax) [inline support/hooks/pre-push:259]: move error message up

File: support/hooks/pre-push [hunk 39]
```diff
@@ -11,83 +11,322 @@
...
+SKIP_SHELLCHECK="${SKIP_SHELLCHECK:-}"
+
+# skip logic for protos - requires bazel
+SKIP_PROTOS="${SKIP_PROTOS:-}"
+BAZEL_PATH=${BAZEL_PATH:-$(command -v bazel)}
+PROTOS_WARNING=(
+    "WARNING: Missing bazel command in \$PATH"
+    ""
+    "  Install bazel for your system:"
+    "      https://github.com/envoyproxy/envoy/blob/main/bazel/README.md#running-clang-format-without-docker")
```
Comment (phlax) [inline support/hooks/pre-push:88]: ~wrong url

File: support/hooks/pre-push [hunk 40]
```diff
@@ -11,83 +11,365 @@
...
+
+check_flake8 () {
+    # Flake8 a python file
+    local file="$1"
+    if [[ -n "$SKIP_FLAKE8" ]]; then
+        return
+    fi
+    log_check flake8 "$file"
+    if [[ -n "$BAZEL_PATH" ]] && [[ -n "$USE_BAZEL_TOOLS" ]];  then
+        "$FLAKE8_PATH" run //tools/code_format:python_flake8 "$file" || handle_error flake8 "$file"
```
Comment (phlax) [inline support/hooks/pre-push:184]: bazel_path

File: support/hooks/README.md [hunk 41]
```diff
@@ -0,0 +1,235 @@
...
+    "protos:\.proto$"
+    "repositories:\.bzl$")
+
+```
+
+See the pre-push file for other settings.
+
+### Missing check utilities
+
+**If a utility required by a check is missing the check will be disabled**, and a warning will be printed to `stderr`.
```
Comment (phlax) [inline support/hooks/README.md:188]: this is incorrect

**Pull request closed without merging**